authors/apps/authentication/views.py

```diff
@@ -223,3 +229,99 @@ def put(self, request, **kwargs):
+
+
+class SocialAuthView(generics.CreateAPIView):
+    queryset = User.objects.all()
+    serializer_class = SocialSignUpSerializer
+    permission_classes = (AllowAny,)
+    renderer_classes = (UserJSONRenderer,)
+
+    def create(self, request, *args, **kwargs):
+        """
+        Override `create` instead of `perform_create` to access request
+        request is necessary for `load_strategy`
+        """
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        return self.create_token(request, serializer)
+
+    def create_token(self, request, serializer):
+        provider = request.data['provider']
+
+        # passing `request` to `load_strategy`,
+        # python-social-auth knows to use the Django strategy.
+        # `strategy` is a PSA concept for referencing Python frameworks
+        # (e.g. Flask, Django, etc.)
+        strategy = load_strategy(request)
+
+        # get backend corresponding to our user's social auth provider
+        # i.e. Google, Facebook, Twitter
+        try:
+            backend = load_backend(
+                strategy=strategy, name=provider, redirect_uri=None)
+        except MissingBackend as e:
+            return Response({
+                "errors": {
+                    "provider": ["Provider not found.", str(e)]
+                }
+
+            }, status=status.HTTP_404_NOT_FOUND)
+
+        if isinstance(backend, BaseOAuth1):
+            try:
+                # Twitter uses OAuth1 and requires that you also pass
+                # an `oauth_token_secret` with your authentication request
+                token = {
+                    'oauth_token': serializer.data['access_token'],
+                    'oauth_token_secret': request.data['access_token_secret'],
+                }
+            except KeyError:
+                return Response({
+                    "errors": "provide access_token and/or access_token_secret"
+
+                }, status=status.HTTP_400_BAD_REQUEST)
+
+        elif isinstance(backend, BaseOAuth2):
+            # only access token required for oauth2
+            token = serializer.data['access_token']
+
+        return self.get_auth_user(request, backend, token, provider)
+
+    def get_auth_user(self, request, backend, token, provider):
+        # If this request is made with an authenticated user,
+        #  try to associate a social account with it
+        authed_user = request.user if not request.user.is_anonymous else None
+
+        try:
+            # if `authed_user` is None,
+            #   python-social-auth will make a new user,
+            # else
+            #   this social account will be associated with the user passed
+            user = backend.do_auth(token, user=authed_user)
+        except AuthAlreadyAssociated:
+            return Response({
+                "errors": "That social media account is already in use"
+            }, status=status.HTTP_400_BAD_REQUEST)
```

Comment:
This should be a conflict status code returned. `409`

Reply:
Done. Changed to 409